uwblueprint/headsupguys
# Add user progress saving

## Notion ticket link
[[Module Experience] Add user progress saving](https://www.notion.so/uwblueprintexecs/Module-Experience-Add-user-progress-saving-48c9ef9320f14fa3acda8e45096a9041)

## Implementation description
When logged in:

- [ ]  Pressing next or prev will send a request to save the progress in the background (should NOT block any UI events)
- [ ]  No save buttons should be rendered

When not logged in:

- [ ]  Pressing save should open the modal to login
- [ ]  Leaving the module with progress open the modal as well
- [ ]  When the modal opens, save the current progress to localstorage (under `unsaved` key, need to save moduleId with the progress)
- [ ]  After the user logs in, if there is unsaved module, save that progress and redirect them back to the module. only if they logged in a from modal do you redirect
- [ ]  Now saving works as though you are logged in


## Steps to test
1. Navigate to [http://localhost:3000/admin/dashboard/modules](http://localhost:3000/admin/dashboard/modules) and create a new module
2. Add at least 4 sections to the first page of the module and fill these sections with content of every type, create a second page and do the same there as well
![image](https://user-images.githubusercontent.com/24593519/167459134-7e9bfb7d-090f-4c6b-88f9-6bd8ae991d3e.png)
3. Check the "save" checkbox at the bottom of the first and second pages
![image](https://user-images.githubusercontent.com/24593519/167461982-1fc30ce2-a8b3-426f-95db-ae54876f9dd8.png)
4. Save your module and take note of the module id that has appeared in your browser's search bar 
![image](https://user-images.githubusercontent.com/24593519/167460015-9afc29d6-84f9-42fe-b32d-bc9ce1b491c5.png)
5. Clear all cookies by clicking the icon on the left hand side of the search bar
![image](https://user-images.githubusercontent.com/24593519/167463353-fea57a5c-f3cd-4e57-88fb-d0ac34c572ef.png)
Clicking on localhost and pressing "remove"
![image](https://user-images.githubusercontent.com/24593519/167463490-d5955057-80b8-491f-9099-10690bc2f7de.png)
6. Navigate to http://localhost:3000/module/<YOUR MODULE ID>
![image](https://user-images.githubusercontent.com/24593519/167460137-6c548415-7845-4096-a756-84db7e51f101.png)
7. Before making any changes, press the back button on your browser, and ensure no pop up appears reminding you to save your changes
8. Go forward in your browser back to your module's page
9. On both pages of your module, click multiple choice and multiselect buttons at random and write data into the text field
10. Now when pressing the back (or reload button) you should be reminded to save your work and prompted to log in. Do not log in yet.
11. Navigate to the second page of your module and press the save button there, you should once again receive a prompt to log in. Click the green Log In Button
12. Sign up and ensure that you are redirected to the same page you left and that your data persists
13. Clear your browser data, fill in the 2 pages at random, and press save again, this time Log in instead of signing up and ensure that you are redirected to the appropriate page and the data is correct
14. Feel free to repeat steps 12 and 13 with the sign up method you did not choose last time, and if you previously filled the demographic survey, do not fill it this time, or vice versa
15. Let me know any bugs or unexpected behaviour that you experience!


## What should reviewers focus on?
* That logging in from the module saves your data and position


## Checklist
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this P
